### PR TITLE
Make LogicalMeterActor resilient to wall-clock time jumps

### PR DESCRIPTION
**Pull request by shsms**
Adds a new internal `WallClockTimer` that schedules resampler ticks against the
wall clock while sleeping on tokio's monotonic clock.

When the two diverge by more than one resampling interval in either direction,
the timer realigns and the actor rebuilds its inner resamplers against the new
clock frame.

Subscribers see a single `None` sample at the resync tick (preserving the
every-interval cadence) and real values resume on the next tick — no catch-up
burst, no duplicate samples.

### CODE REVIEW

File: src/wall_clock_timer.rs [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+fn align_to_epoch(timestamp: DateTime<Utc>, interval: TimeDelta) -> DateTime<Utc> {
+    let interval_nanos = interval.num_nanoseconds().unwrap_or(0);
+    if interval_nanos <= 0 {
+        return timestamp;
+    }
+    let ts_nanos = timestamp
+        .signed_duration_since(DateTime::<Utc>::UNIX_EPOCH)
+        .num_nanoseconds()
+        .unwrap_or(0);
+    let aligned = (ts_nanos / interval_nanos) * interval_nanos;
```
[the remark this answers is not in the document]
Comment (shsms) [reply src/wall_clock_timer.rs]: Done

File: src/wall_clock_timer.rs [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+/// negative.
+fn align_to_epoch(timestamp: DateTime<Utc>, interval: TimeDelta) -> DateTime<Utc> {
+    let interval_nanos = interval.num_nanoseconds().unwrap_or(0);
+    if interval_nanos <= 0 {
+        return timestamp;
+    }
+    let ts_nanos = timestamp
+        .signed_duration_since(DateTime::<Utc>::UNIX_EPOCH)
+        .num_nanoseconds()
+        .unwrap_or(0);
```
[the remark this answers is not in the document]
Comment (shsms) [reply src/wall_clock_timer.rs]: Done

File: src/logical_meter/logical_meter_actor.rs [hunk 3]
```diff
@@ -244,47 +264,25 @@ impl Formulas {
...
-            (last_aligned_ts + config.resampling_interval - now)
-                .to_std()
-                .map_err(|e| Error::chrono_error(format!("Failed to calculate time delta: {e}")))?,
-        );
+        let timer = WallClockTimer::new(config.resampling_interval, clock);
+        // Resamplers created before the first tick use `resampler_ts` as
+        // their start; setting it one interval before the first scheduled
+        // tick lines up with the original semantics (first tick produces the
+        // first resampled sample).
+        let resampler_ts = timer.next_tick_time() - config.resampling_interval;
```
[the remark this answers is not in the document]
Comment (shsms) [reply src/logical_meter/logical_meter_actor.rs]: done

File: src/wall_clock_timer.rs [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+    #[tokio::test]
+    async fn test_timer_ticks_at_interval() {
+        let interval = TimeDelta::try_milliseconds(20).unwrap();
+        let mut timer = WallClockTimer::new(interval, SystemClock);
+        let t1 = timer.tick().await.expected_tick_time;
+        let t2 = timer.tick().await.expected_tick_time;
+        let t3 = timer.tick().await.expected_tick_time;
+        assert_eq!(t2 - t1, interval);
+        assert_eq!(t3 - t2, interval);
+    }
```
[the remark this answers is not in the document]
Comment (shsms) [reply src/wall_clock_timer.rs]: fixed.

**Comment by hannah-stevenson-frequenz** [comment]
my understanding, which might still be wrong, is that the problem is:

samples are wall clock stamped. i'm not sure if it's when they were collected or when they arrive but this is all on the same machine so there won't be much difference
the resampler has a start variable, set at instantiation time.
when the actor asks for a sample it passes in resampler_ts which is a free running clock, updated by a fixed amount every time the monotonic clock says a time has past but disconnected from the wall clock.
the sample window is [start, resampler_ts)
when the wall clock has drifted too much then the sampling window no longer reflects the past resample period.

it might be simpler if the resampler is asked for the "last sample" which is the current wall clock time, floored to some point (e.g. whole seconds) to make sure different clients are aligned, minus the resampling interval and no clock is kept at all. the client wakes up based on the monotonic clock, asks for the last sample and it doesn't matter about shifting clocks. whilst simpler there maybe use cases this would break for of which i am unaware - i don't know if ppl try to get historical data for example, or multiple samples. this isn't the approach of the PR though.

this is just for other confused readers as it took me a while to get my head around this and the problem has popped up a few times.

**Comment by llucax** [comment]
> it might be simpler if the resampler is asked for the "last sample" which is the current wall clock time, floored to some point (e.g. whole seconds) to make sure different clients are aligned, minus the resampling interval and no clock is kept at all.

Interesting. Not sure about the resampler in this repo, but on the SDK (this is based on the SDK timer AFAIK) that's not good enough because you could stop receiving data, so asking for the last sample might give you a very old sample (and the resampler will still send resampler data anyway, worse case with value = None, but we want it to keep ticking), so not very reliable for that scenario.

**Comment by hannah-stevenson-frequenz** [comment]
>so asking for the last sample might give you a very old sample

it wouldn't because it would be outside of the time window. it wouldn't go backs searching for data, just things it already has within the window.

File: src/wall_clock_timer.rs [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+        loop {
+            let now = self.clock.wall_now();
+            let to_next = self.next_tick - now;
+            let drift = if slept_this_call {
+                to_next
+            } else {
+                to_next - self.interval
+            };
+
+            if drift.abs() > threshold {
```
Comment (llucax) [inline src/wall_clock_timer.rs]: This jump detector will also fire when the caller is simply late to call `tick()` again, even if the wall clock never jumped.

On entry (`slept_this_call == false`) we compute `drift = to_next - self.interval`. That assumes `tick()` is called almost immediately after the previous tick, so `to_next ~= interval`. But `LogicalMeterActor::run()` does real work between ticks (`resample_metrics()`, formula evaluation, broadcast sends, scheduling delays, etc.). If that work takes more than one extra interval, `to_next` is already negative by the time we enter the next `tick()`, so `drift.abs() > threshold` becomes true and we resync even though this is just a slow consumer / event-loop delay.

That is a correctness regression relative to the Python implementation: there the drift check is based on `wall_clock_elapsed - monotonic_elapsed` around each sleep, so a late caller does not look like a wall-clock jump.

Impact: under load this can emit spurious `resynced` ticks, rebuild all inner resamplers, drop buffered telemetry, and surface an unexpected `None` sample to subscribers.

Please switch the jump detection to compare wall-clock vs monotonic elapsed time across the sleep (or otherwise exclude normal catch-up lateness on entry). A regression test that advances tokio time by >1 interval between two `tick()` calls without calling `inject_wall_jump()` would catch this.

File: src/wall_clock_timer.rs [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+impl<C: Clock> WallClockTimer<C> {
+    pub fn new(interval: TimeDelta, clock: C) -> Self {
+        let now = clock.wall_now();
+        let next_tick = align_to_epoch(now, interval) + interval;
+        Self {
+            interval,
+            next_tick,
+            clock,
+        }
+    }
```
Comment (llucax) [inline src/wall_clock_timer.rs]: `LogicalMeterActor::try_new()` now rejects non-positive `resampling_interval`, but the timer constructor itself still accepts `interval <= 0`.

That leaves the new public timer type in an invalid state: `align_to_epoch()` falls back to returning the input timestamp for zero/negative intervals, `next_tick` is initialized to `now + interval`, and `tick()` can then either return immediately forever or otherwise behave nonsensically instead of representing a real periodic schedule.

The Python implementation validates this upfront and raises on non-positive intervals. Please do the same here (ideally by making `WallClockTimer::new()` return a `Result`, or at least asserting and documenting the precondition), and add direct constructor coverage for `0` and negative intervals.

File: RELEASE_NOTES.md [hunk 7]
```diff
@@ -36,6 +36,8 @@
   - `power()` — a `Formula<Power>` for the pool's aggregated power.
   - `power_bounds()` — a `broadcast::Receiver<Vec<Bounds<Power>>>` tracking the pool's power bounds.
 
+- The logical meter now survives NTP-style wall-clock jumps. A new internal `WallClockTimer` schedules resampler ticks against the wall clock while sleeping on tokio's monotonic clock; when the two diverge by more than one interval in either direction, the timer realigns and the actor rebuilds its inner resamplers against the new clock frame. Subscribers see one `None` sample at the resync tick (preserving the every-interval cadence) and real values resume on the next tick.
```
Comment (llucax) [inline RELEASE_NOTES.md]: This release note covers the new behavior, but the PR also contains upgrade-significant API changes that are still missing here:

1. `ErrorKind::ChronoError` is removed from the public enum, so downstream exhaustive matches will stop compiling after upgrade.
2. `InvalidConfig` is added and non-positive `resampling_interval` now fails construction instead of being accepted.

Those should be called out explicitly in `Upgrading` (or at least under breaking changes / bug fixes), otherwise users only reading the release notes won't know why a previously compiling match or config now fails.

File: src/error.rs [hunk 8]
```diff
@@ -56,10 +56,10 @@ ErrorKind!(
     (ComponentGraphError, component_graph_error),
     (ComponentDataError, component_data_error),
     (ConnectionFailure, connection_failure),
-    (ChronoError, chrono_error),
     (DroppedUnusedFormulas, dropped_unused_formulas),
     (FormulaEngineError, formula_engine_error),
     (InvalidComponent, invalid_component),
+    (InvalidConfig, invalid_config),
```
Comment (llucax) [inline src/error.rs:62]: Removing `ErrorKind::ChronoError` is a breaking public API change: downstream code that matches exhaustively on `ErrorKind` will stop compiling after this lands.

If that break is intentional, it should be called out explicitly in the release notes / upgrading notes. If it is not intentional, the safer option is to keep the enum variant for now even if no code constructs it anymore, and remove it only in the next breaking release.

File: src/wall_clock_timer.rs [hunk 9]
```diff
@@ -0,0 +1,391 @@
+// License: MIT
+// Copyright © 2026 Frequenz Energy-as-a-Service GmbH
+
+//! A minimal wall-clock-aligned timer that survives NTP jumps.
+//!
+//! Sleeps run on tokio's monotonic clock; tick times are expressed on the
+//! wall clock. If the two diverge by more than one interval (detected at
+//! tick entry), the timer realigns to the wall clock and the next tick
+//! fires immediately with `TickInfo::resynced = true`.
```
Comment (llucax) [inline src/wall_clock_timer.rs:9]: Compared with the Python implementation this timer still has a much narrower contract: it is always epoch-aligned, it does not expose `align_to`, it does not track gradual wall-clock drift, and it does not warn on async scheduling drift.

That may be enough for the current `LogicalMeterActor` use case, but it means the Rust version is not yet feature-parity with `frequenz-sdk-python`'s `WallClockTimer`. Since this PR and the release note describe it as being based on that implementation, I’d either document the intentionally reduced scope here, or keep porting the missing reliability features so future reuse doesn’t assume parity that isn’t there yet.

**Comment by llucax** [comment]
> it wouldn't because it would be outside of the time window. it wouldn't go backs searching for data, just things it already has within the window.

So if you don't consider old data, how do you know if there is no data because you stopped receiving data or because  you jumped into the future? How do you decide which timestamp to use to emit those samples to keep the continuous ticking?

I have the feeling is not that simple, but again, I went deep into this a long time ago so I don't have it fresh. Since I was already asking AI to do a PR review, I asked it to help me consider this simple approach. The reply is long, but it seems to align with my intuition that for continuous aligned ticking it might not be that trivial. 

<details>
<summary>AI analysis</summary>

That proposal makes sense as a description of **why time jumps break the current resampling windows**, but it does **not** look like a complete replacement for the PR/SDK approach.

**Short version**

The proposed "just wake up on the monotonic clock and resample the latest wall-clock bucket" can work for a **different API**, but not cleanly for the current one, because the current logical meter needs a **stream with one tick per interval**, aligned to wall-clock boundaries as closely as possible, including when there is **no input data**.

**Why the simpler idea is not a drop-in replacement**

1. It still needs state.

If you compute `bucket = floor(wall_now, interval)` on every wake-up, you still need to remember what bucket you emitted last.

Without that:
- if you wake twice in the same bucket, you duplicate output
- if you wake late and skip buckets, you silently drop intervals
- if the wall clock jumps backwards, you may ask for an already-emitted bucket again

So "no clock is kept at all" is not really true once you define correct behavior.

2. Backward jumps are the hardest case, and this proposal does not solve them.

Example:
- you already emitted the bucket ending at `12:00:10`
- wall clock jumps back to `11:59:40`
- next monotonic wake computes a much earlier bucket

Now what?

Possible behaviors are all problematic:
- emit that older bucket again: duplicates / time goes backwards
- refuse to emit until wall clock catches back up: you get a long gap with no ticks
- rebuild and realign immediately: that is basically what this PR does

So clock shifts do still matter; this proposal just postpones the policy decision.

3. Forward jumps or slow processing can skip intervals.

If wall clock jumps forward by 30 seconds, or the actor is blocked for 30 seconds, then "give me the latest completed bucket" only gives you one bucket.

That means you either:
- drop 29 intervals, which breaks the current streaming semantics
- or emit all missed intervals, which again requires remembered state and catch-up logic

The current code expects exactly one resampled output per tick. You can see that in `src/logical_meter/logical_meter_actor.rs:511-517`.

4. It does not give wall-clock-aligned ticks, only wall-clock-aligned timestamps.

Those are different.

With the proposal, you wake on a monotonic cadence and only then ask "what was the latest completed wall-clock bucket?". That means the sample may be produced noticeably after the bucket boundary.

The current timer tries to do better: it sleeps on monotonic time but schedules ticks against wall-clock boundaries.

So if the requirement is "tick every interval, as aligned to the wall clock as possible", the simpler proposal is weaker.

5. It does not match the current resampler model very well.

The current resampler is stateful. It has a start/alignment and is advanced over time. The actor rebuilds it after a jump because that internal alignment frame becomes stale.

So the simpler idea would fit better with a different resampler API like:
- "given `[start, end)`, compute that bucket from retained history"

But that is not the current design. With the current design, backward jumps and re-asking old buckets are especially awkward.

6. Missing-data periods still need regular ticks.

Your intuition here is correct.

If no telemetry arrives for a while, the logical meter still needs to emit one output per interval, typically `None`.

A "latest bucket" query can do that only if you are still waking reliably once per interval and tracking which buckets have already been emitted. Again, that brings back timer/state logic.

**What the proposal gets right**

The proposal is right about the core bug mechanism:

- telemetry samples are stamped in wall-clock time
- resampling windows must therefore also be defined in wall-clock time
- if your output schedule drifts away from the wall-clock frame, windows become wrong

That part is valid.

**What it misses**

It misses that the problem is not only "which bucket is latest now?", but also:

- exactly-once emission per interval
- behavior on backward jumps
- behavior when several intervals are crossed between wake-ups
- emitting `None` during silent periods
- keeping multiple subscribers aligned

Those are exactly the reasons the PR/SDK use a wall-clock-aligned timer plus explicit resync behavior.

**Bottom line**

The proposal is feasible only if you accept a different contract:

- "give me the latest completed bucket when I happen to wake up"

It is not equivalent to:

- "produce a continuous per-interval stream aligned to wall-clock boundaries, including through time jumps and missing-data periods"

For the current logical meter semantics, the simpler proposal has real flaws, and your intuition is not wrong.

</details>

File: src/wall_clock_timer.rs [hunk 10]
```diff
@@ -0,0 +1,391 @@
...
+        //   - At entry (`slept_this_call == false`), `to_next` is naturally
+        //     ~= `interval` — the previous tick advanced `next_tick` by one
+        //     interval and we've barely moved since. Compare `to_next -
+        //     interval` against the threshold, so the expected value is 0.
+        //
+        //   - After a sleep, `to_next` is naturally ~= 0 (we slept exactly
+        //     until the scheduled tick). Compare `to_next` directly.
+        // Strict `>`: drift of exactly 1× interval is treated as scheduling
+        // jitter, not a jump. Only strictly-greater drifts resync.
+        let threshold = self.interval;
```
Comment (llucax) [inline src/wall_clock_timer.rs:134]: The comment about the threshold being `1 × interval` is not quite what the current math implements on entry.

After a previous tick, the first iteration uses `drift = to_next - self.interval`. That means an entry-time forward shift has to exceed roughly `2 × interval` before `drift.abs() > threshold` fires, while a late caller can trip the opposite side much earlier. So the code and the prose have drifted apart here.

I’d either update the implementation so the threshold really is symmetric at `1 × interval`, or adjust the comment/release notes to describe the current behavior precisely.

File: src/wall_clock_timer.rs [hunk 11]
```diff
@@ -0,0 +1,391 @@
...
+            if to_next <= TimeDelta::zero() {
+                let expected = self.next_tick;
+                self.next_tick = expected + self.interval;
+                return TickInfo {
+                    expected_tick_time: expected,
+                    resynced: false,
+                };
+            }
+
+            tokio::time::sleep(to_next.to_std().unwrap_or(std::time::Duration::ZERO)).await;
```
Comment (llucax) [inline src/wall_clock_timer.rs]: This `unwrap_or(Duration::ZERO)` turns an unrepresentable `TimeDelta` into a zero-duration sleep, which means a misconfigured large interval can become a hot loop instead of a clear error.

That is especially awkward now that `LogicalMeterActor::try_new()` already validates some timer preconditions. It is probably much safer to validate here as well (or returning an error from `new()` / `tick()`) so unsupported intervals fail deterministically instead of spinning the executor.

File: src/wall_clock_timer.rs [hunk 12]
```diff
@@ -0,0 +1,391 @@
...
+        let now = clock.wall_now();
+        let next_tick = align_to_epoch(now, interval) + interval;
+        Self {
+            interval,
+            next_tick,
+            clock,
+        }
+    }
+
+    pub fn next_tick_time(&self) -> DateTime<Utc> {
```
Comment (llucax) [inline src/wall_clock_timer.rs:117]: It would help to document here that `next_tick_time()` is expressed on the wall clock and can move backwards after a backward jump / resync.

That behavior is expected, but it is easy for callers to assume successive values are monotonic and use them for ordering or latency calculations.

File: src/wall_clock_timer.rs [hunk 13]
```diff
@@ -0,0 +1,391 @@
...
+use chrono::{DateTime, TimeDelta, Utc};
+
+/// Abstracts wall-clock time so tests can wire it to tokio's paused clock.
+/// Production uses [`SystemClock`]; test-only implementations live in
+/// `crate::client::test_utils`.
+pub trait Clock: Send + Sync {
+    fn wall_now(&self) -> DateTime<Utc>;
+}
+
+/// The real system wall clock. `wall_now()` returns `chrono::Utc::now()`.
```
Comment (llucax) [inline src/wall_clock_timer.rs:27]: It might be worth documenting one subtlety here: `SystemClock` is fine in production, but it is the wrong choice for `#[tokio::test(start_paused = true)]` tests because wall time keeps coming from the real system clock while tokio's monotonic time is frozen.

The new `TokioSyncedClock` solves that nicely; a short doc note on `SystemClock` would make that testability constraint much more discoverable.

File: src/client/test_utils/tokio_synced_clock.rs [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+                mono_anchor: Instant::now(),
+            })),
+        }
+    }
+
+    /// Shifts wall-clock time by `offset` relative to the monotonic clock,
+    /// simulating an NTP jump. Visible to every clone.
+    pub fn inject_wall_jump(&self, offset: TimeDelta) {
+        let mut inner = self.inner.write().expect("clock poisoned");
+        inner.wall_anchor += offset;
```
Comment (llucax) [inline src/client/test_utils/tokio_synced_clock.rs:65]: Nit: A short inline note would help preserve the invariant here that only `wall_anchor` changes and `mono_anchor` intentionally does *not*.

That detail is what makes this simulate an NTP jump instead of a full re-anchor of both clocks, and it is easy to accidentally "simplify" in a future refactor.

File: src/logical_meter/logical_meter_actor.rs [hunk 15]
```diff
@@ -499,6 +524,39 @@ impl LogicalMeterActor {
...
+        &self,
+        resamplers: &mut HashMap<(u64, Metric), ComponentDataResampler>,
+        start: DateTime<Utc>,
+    ) {
+        for resampler in resamplers.values_mut() {
+            // Drain any samples that were queued during the jump window;
+            // they are timestamped on the old wall-clock frame and would
+            // pollute the freshly-aligned resampler.
+            while poll_telemetry(&mut resampler.receiver, resampler.component_id).is_some() {}
+            let function = self
```
Comment (llucax) [inline src/logical_meter/logical_meter_actor.rs]: This rebuild path duplicates the resampling-function selection logic from `start_resamplers()` above, and it also duplicates the `max_age_in_intervals` clamp a few lines later.

Now that the jump path is correctness-critical, I’d extract a small helper for "build resampler config for metric X" so startup and rebuild cannot silently diverge as `LogicalMeterConfig` evolves.

File: src/logical_meter/logical_meter_actor.rs [hunk 16]
```diff
@@ -34,6 +34,26 @@ struct ComponentDataResampler {
...
+fn poll_telemetry(
+    receiver: &mut broadcast::Receiver<ElectricalComponentTelemetry>,
+    component_id: u64,
+) -> Option<ElectricalComponentTelemetry> {
+    loop {
+        match receiver.try_recv() {
+            Ok(data) => return Some(data),
+            Err(tokio::sync::broadcast::error::TryRecvError::Empty) => return None,
+            Err(tokio::sync::broadcast::error::TryRecvError::Lagged(n)) => {
+                tracing::debug!("resampler receiver lagged {n} samples for cid={component_id}");
```
Comment (llucax) [inline src/logical_meter/logical_meter_actor.rs]: When this happens during jump recovery we are dropping telemetry, not just handling an internal scheduling detail. Logging it only at `debug` makes the data-loss path easy to miss operationally.

I’d consider `info` or `warn` here, at least when this helper is being used from the jump-drain path, so operators can tell that samples were discarded across the wall-clock jump.

File: src/logical_meter/logical_meter_actor.rs [hunk 17]
```diff
@@ -499,6 +524,39 @@ impl LogicalMeterActor {
...
+                .config
+                .resampling_overrides
+                .get(&resampler.metric)
+                .cloned()
+                .or_else(|| self.config.resampling_function.clone())
+                .unwrap_or(ResamplingFunction::Average);
+            resampler.resampler = frequenz_resampling::Resampler::new(
+                self.config.resampling_interval,
+                function,
+                self.config.max_age_in_intervals.min(i32::MAX as u32) as i32,
```
Comment (llucax) [inline src/logical_meter/logical_meter_actor.rs]: This silently truncates user configuration to `i32::MAX`.

If someone passes a very large `max_age_in_intervals`, they will get a different effective retention window with no feedback. I’d prefer validating it once in `LogicalMeterConfig` / actor construction and surfacing an `InvalidConfig` error instead of quietly clamping it here (and in the startup path).

File: src/client/test_utils.rs [hunk 18]
```diff
@@ -240,58 +247,54 @@ impl MockComponent {
...
             .unwrap_or_default();
-        let next_sec =
-            SystemTime::UNIX_EPOCH + std::time::Duration::from_secs(since_epoch.as_secs() + 1);
-
-        // Sleep until the start of the next second to align telemetry
-        // timestamps with the resampler's clock, so that we get reproducible
-        // values from the logical meter.
+        let next_sec_secs = since_epoch.as_secs() + 1;
+        let next_sec = SystemTime::UNIX_EPOCH + std::time::Duration::from_secs(next_sec_secs);
         std::thread::sleep(next_sec.duration_since(now).unwrap_or_default());
```
Comment (llucax) [inline src/client/test_utils.rs]: This still introduces up to ~1 second of real wall-clock delay into tests that go through `MockMicrogridApiClient::new()`, even though this PR added `new_with_clock()` specifically to keep the new timer tests deterministic.

It may be worth migrating more internal tests over to the clock-injected constructor, or at least documenting that `new()` is the slower / real-time path.

File: src/wall_clock_timer.rs [hunk 19]
```diff
@@ -0,0 +1,391 @@
...
+
+        clock.inject_wall_jump(-TimeDelta::try_milliseconds(500).unwrap());
+        let info = timer.tick().await;
+        assert!(!info.resynced);
+    }
+
+    // The threshold is `1 × interval`: a jump just above the interval
+    // should resync.
+    #[tokio::test(start_paused = true)]
+    async fn test_timer_detects_forward_jump_just_over_interval() {
```
Comment (llucax) [inline src/wall_clock_timer.rs:382]: These `just over interval` tests are useful, but there is still no coverage for the exact boundary case (`== 1 × interval`).

That boundary matters because both the implementation and the comments rely on the distinction between `>` and `>=`. I’d add exact-threshold forward and backward cases so the intended resync boundary stays pinned.

File: src/client/test_utils.rs
Comment (llucax) [inline src/client/test_utils.rs:1]: Super minor and I guess it is not meant to be public interface, but this refactor also changes the public test-utils surface by removing `MockComponent::with_start_times` and switching timestamp generation over to the shared clock. If any downstream tests use these public helpers it is a breaking change worth documenting.

File: src/logical_meter/logical_meter_actor.rs [hunk 20]
```diff
@@ -499,6 +524,39 @@ impl LogicalMeterActor {
...
+            // pollute the freshly-aligned resampler.
+            while poll_telemetry(&mut resampler.receiver, resampler.component_id).is_some() {}
+            let function = self
+                .config
+                .resampling_overrides
+                .get(&resampler.metric)
+                .cloned()
+                .or_else(|| self.config.resampling_function.clone())
+                .unwrap_or(ResamplingFunction::Average);
+            resampler.resampler = frequenz_resampling::Resampler::new(
```
Comment (llucax) [inline src/logical_meter/logical_meter_actor.rs]: The rebuild path depends on a pretty specific contract from `frequenz_resampling::Resampler`: after rebuilding with `start = current - interval`, the very next resample at the realigned tick must still yield exactly one output, even with an empty buffer, so subscribers get the single `None` sample described in the PR.

If that upstream contract ever changes (for example, zero outputs for an empty window), this code flips from graceful recovery into `ConnectionFailure("Resampling produced N values")` at runtime. I’d pin that assumption with a focused regression test around rebuild-after-jump behavior, not only the broader end-to-end recovery test.

**Review by llucax — COMMENTED** [review]
OK, this is a first *AI-assisted* round (I asked it to compare to the SDK implementation, and did a quick check on the findings). To be honest, to do an in-depth round myself will probably take too much time, as it's been a while since I did the SDK implementation and I'm not familiar with rust. We can see if it is necessary after this round.

Some extra comments about commits:

1. `f7e9d09` (*Add NTP-jump-resilient wall clock timer*) and `abfa4b4` (*Use WallClockTimer in LogicalMeterActor for time-jump resilience*) are both substantial functional commits, but both have empty commit bodies. Given how much behavior changed here, a short explanation of the design tradeoffs in the commit messages would make the history much easier to follow later.
2. `56ebc5e` (*Remove ChronoError variant from ErrorKind*) removes the public `ErrorKind::ChronoError` variant, but the commit message does not explain why that removal is safe / intended. Since that is upgrade-significant, the rationale should ideally be visible in history too.

**Comment by hannah-stevenson-frequenz** [comment]
> So if you don't consider old data, how do you know if there is no data because you stopped receiving data or because you jumped into the future?

There is no simple way. I did say there may be hidden requirements I was unaware of - this just gets you samples from the last sample period and not much else. If we want to monitor the stream and do some stats and e.g. give reasons there were no samples in a window I would probably have that separately - but it would depend on exactly what monitoring we wanted to perform.

> How do you decide which timestamp to use to emit those samples to keep the continuous ticking?

I'm not sure what is the continuous ticking that you want? I was just solving the problem of "samples in the last window". If there are other requirements I am unaware of them and they won't be taken into account. This probably isn't the place for this discussion so if you want to continue it we should write out the requirements and then see if there is anything simpler.

On the AI comments, most assume more requirements than the last window - the ability to fetch historical windows for example. That could be added but again without clear requirements thinking about it is fruitless.

I added the comment because the PR was hard to understand (for me) given I didn't understand the design, and didn't understand the reason for the design, and I thought others might be in the same boat.

File: src/wall_clock_timer.rs (hunk 5 above)
Comment (shsms) [reply src/wall_clock_timer.rs]: Damn, that was a real bug, that I introduced because I asked it to simplify.  Switched back to the SDK impl.

File: src/wall_clock_timer.rs (hunk 6 above)
Comment (shsms) [reply src/wall_clock_timer.rs]: Fixed.

File: RELEASE_NOTES.md (hunk 7 above)
Comment (shsms) [reply RELEASE_NOTES.md]: Updated.

File: src/error.rs (hunk 8 above)
Comment (shsms) [reply src/error.rs:62]: None of our downstreams are doing an exhaustive match.  For others it is very easy to migrate.

File: src/wall_clock_timer.rs (hunk 9 above)
Comment (shsms) [reply src/wall_clock_timer.rs:9]: removed references to the SDK and just described what has been implemented.

File: src/wall_clock_timer.rs (hunk 10 above)
Comment (shsms) [reply src/wall_clock_timer.rs:134]: The implementation has been replaced with drift between monotonic and wall clocks.

File: src/wall_clock_timer.rs (hunk 11 above)
Comment (shsms) [reply src/wall_clock_timer.rs]: if to_next doesn't fit in a std::duration (which can only happen if there's a time jump of 100s of years), we're sleeping for the interval instead.

File: src/wall_clock_timer.rs (hunk 12 above)
Comment (shsms) [reply src/wall_clock_timer.rs:117]: done.

File: src/wall_clock_timer.rs (hunk 13 above)
Comment (shsms) [reply src/wall_clock_timer.rs:27]: done

File: src/client/test_utils/tokio_synced_clock.rs (hunk 14 above)
Comment (shsms) [reply src/client/test_utils/tokio_synced_clock.rs:65]: done.

File: src/logical_meter/logical_meter_actor.rs (hunk 15 above)
Comment (shsms) [reply src/logical_meter/logical_meter_actor.rs]: done, there's now a `build_resampler` method called from both places.

File: src/logical_meter/logical_meter_actor.rs (hunk 16 above)
Comment (shsms) [reply src/logical_meter/logical_meter_actor.rs]: switched to warn.

File: src/logical_meter/logical_meter_actor.rs (hunk 17 above)
Comment (shsms) [reply src/logical_meter/logical_meter_actor.rs]: this is fine, they better not send much bigger values.

File: src/client/test_utils.rs (hunk 18 above)
Comment (shsms) [reply src/client/test_utils.rs]: Fixed by removing the sleep and using the same TokioSyncedClock (anchored to an epoch-aligned ts) everywhere.

File: src/wall_clock_timer.rs (hunk 19 above)
Comment (shsms) [reply src/wall_clock_timer.rs:382]: fixed.

File: src/client/test_utils.rs
Comment (shsms) [reply src/client/test_utils.rs:1]: There aren't any yet.

File: src/logical_meter/logical_meter_actor.rs (hunk 20 above)
Comment (shsms) [reply src/logical_meter/logical_meter_actor.rs]: done.

**Review by llucax — APPROVED** [review]
Approving from the logical part, not sure if anyone else with more rust-fu should approve too, I leave that to you to decide.
